### PR TITLE
Generate System Info Report

### PR DESCRIPTION
**Pull request by Thelolguy1**
This should in theory reduce the workload on helpdesk.

I am so sick of asking "what cpu do you have"....

### CODE REVIEW

File: cmd/vinegar/vinegar.go [hunk 1]
```diff
@@ -91,6 +92,13 @@ func main() {
 			if err := InstallWebview2(&pfx); err != nil {
 				log.Fatal(err)
 			}
+		case "reportinfo":
+			report, err := util.GenerateInfo(*configPath)
```
Comment (apprehensions) [inline cmd/vinegar/vinegar.go:96]: Pass the configuration struct, not the file. As the struct member will just be the contents of the file.

File: cmd/vinegar/vinegar.go [hunk 2]
```diff
@@ -91,6 +92,13 @@ func main() {
...
+			if err != nil {
+				log.Fatal(err)
+			}
+			fmt.Println("Please share the information below.")
```
Comment (apprehensions) [inline cmd/vinegar/vinegar.go:100]: Remove, using this subcommand already implies it.

File: util/reportinfo.go [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+
+func GenerateInfo(currentConfigurationPath string) (SysInfo, error) {
+	//TODO, returns struct Sysinfo.
+	currentSystem := &SysInfo{}
+
+	// Check for AVX
+	if cpufile, err := os.ReadFile("/proc/cpuinfo"); err != nil {
+		return SysInfo{}, err
+	} else {
+		exp := regexp.MustCompile(`avx`)
```
Comment (apprehensions) [inline util/reportinfo.go:36]: Parse all available CPU flags in string array, and checking for if some specific CPU flags exist, such as avx, avx2, etc.

File: util/reportinfo.go [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+		matches := exp.FindStringSubmatch(string(cpufile))
+		currentSystem.AVXAvailable = (len(matches) > 0)
+	}
+
+	// Get Distro
+	if distro, err := os.ReadFile("/etc/os-release"); err != nil {
+		return SysInfo{}, err
+	} else {
+		currentSystem.Distro = string(distro)
```
Comment (apprehensions) [inline util/reportinfo.go:45]: This file has a specific format, please format as necessary.

File: util/reportinfo.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+		return SysInfo{}, err
+	} else {
+		currentSystem.Config = string(config)
+	}
+
+	// Check if in flatpak
+	if _, err := os.Stat("/.flatpak-info"); err == nil {
+		currentSystem.InFlatpak = true
+	} else if errors.Is(err, os.ErrNotExist) {
+		currentSystem.InFlatpak = false
```
Comment (apprehensions) [inline util/reportinfo.go:66]: This will already be false.

File: util/reportinfo.go [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+	AVXAvailable bool
+	Distro       string //Done
+	Kernel       string // Done
+	InFlatpak    bool   // Done
+	Config       string // Done
+}
+
+func GenerateInfo(currentConfigurationPath string) (SysInfo, error) {
+	//TODO, returns struct Sysinfo.
+	currentSystem := &SysInfo{}
```
Comment (apprehensions) [inline util/reportinfo.go:30]: Why are you using a pointer?

**Review by apprehensions — CHANGES_REQUESTED** [review]
Overall, the error checking is bland, and variable names or code style should be inlign with the Go code style.

Try seeing if there is a library that can provide these sort of features.

File: cmd/vinegar/vinegar.go [hunk 7]
```diff
@@ -91,6 +92,12 @@ func main() {
 			if err := InstallWebview2(&pfx); err != nil {
 				log.Fatal(err)
 			}
+		case "reportinfo":
+			report, err := extra.GenerateInfo(cfg)
```
Comment (apprehensions) [inline cmd/vinegar/vinegar.go:96]: Pass configuration as pointer to prevent more memory usage or garbage collection

**Review by apprehensions — CHANGES_REQUESTED** [review]
Extract all else to outside the function if clause

File: cmd/vinegar/vinegar.go (hunk 7 above)
Comment (Thelolguy1) [reply cmd/vinegar/vinegar.go:96]: It's already a pointer

Comment (Thelolguy1) [reply cmd/vinegar/vinegar.go:96]: i lied

**Comment by Noted-Jrelvas** [comment]
Excluding feedback already given and the build error, things are looking good so far.

Not sure if this is going to be included in this PR, but here's some ideas of information we could be including in the future:
- Vinegar version
- Last Wine log
- Output from vulkaninfo(?) - Would be very useful for identifying driver and hardware incompatibilities

**Comment by Thelolguy1** [comment]
Merged with wael's branch

**Comment by apprehensions** [comment]
> * Vinegar version

Unfortunately we had dropped this due to a conflict on how versioning should be handled.

> * Last Wine log

This is definitely possible, but to as this only gains system information, doesn't seem worth it.

> * Output vulkaninfo

No

**Pull request closed without merging**